distributed/deploy/spec.py

```diff
@@ -11,6 +11,48 @@
 from ..security import Security
 
 
+class ProcessInterface:
+    """ An interface for Scheduler and Worker processes for use in SpecCluster
+
+    Parameters
+    ----------
+    loop:
+        A pointer to the running loop.
+
+    """
+
+    def __init__(self, loop=None):
+        self.address = None
+        self.loop = loop
+        self.lock = asyncio.Lock()
+        self.status = "created"
+
+    def __await__(self):
+        async def _():
+            async with self.lock:
+                if self.status == "created":
+                    await self.start()
+                    assert self.status == "running"
+            return self
+
+        return _().__await__()
+
+    async def start(self):
+        """ Start the process. """
+        self.status = "running"
+
+    async def close(self):
+        """ Close the process. """
+        self.status = "closed"
+
+    async def logs(self):
+        """ Generator which yeilds log messages. """
+        raise NotImplementedError()
```

Comment:
I suspect that in other resource managers we won't get a stream of logs, we'll just get a dump of the logs such as they are right now.  

- @jacobtomlinson What is the situation in Kubernetes?
- @jcrist What is the situation in Yarn?

Reply:
Also https://github.com/dask/dask-jobqueue/issues/299

Reply:
Actually, I probably should have raised this conversation at https://github.com/dask/distributed/pull/2875

Perhaps people can respond there instead.

Reply:
Currently in kubernetes you get a dump. However when moving to async kubernetes we will get a stream. 

We can always encourage folks to present a dump as a stream for consistency anyway.

```python
for line in dump.split('\n'):
    yield line
```

Reply:
From #2875 we should do something better than raise a `NotImplementedError` like get logs back from the scheduler.

In the case that this is the scheduler we can open a `rpc` to `self.address` and call `logs()`. In the case that this is a worker we also need the same rpc, then call `worker_logs` and figure out which is ours.

Reply:
Another option would be to put this logic up in the `SpecCluster` where we catch the `NotImplementedError` and gather these details curectly from the `cluster.scheduler_comm` up there.

Reply:
There's no public api for getting logs in yarn besides the yarn CLI (which implements all of it in the CLI instead of a public method :/). Currently users can get running logs through skein's Web UI (click on the app in the yarn web ui and you get taken here), or use `yarn logs -applicationId appid` to get logs after completion.

I likely could figure out how to get logs as a method and add that to Skein if desired. Besides a few setup scripts though the logs should be equivalent to using the `get_scheduler_logs`/`get_worker_logs` methods.

Reply:
Currently `get_scheduler_logs`/`get_worker_logs` only grab output from the `distributed` Python logger, and nothing else.  I suspect that Yarn captures everything from stdout/stderr, which is a bit nicer.  In general I think that having a uniform way for users to get logs would have some value.  If it's easy for us to do and we can coordinate some way on how to do it cleanly then I think we should do it.

Reply:
I agree, I'm just saying that having a default method that uses the rpc for now (as @jacobtomlinson suggested) would be good, and other implementations can override with their own implementations where possible.

Reply:
Ah great.  It sounds like we're in agreement.  I'll draft something up and submit a PR for general review by the deployment folks sometime hopefully today.

Reply:
In the mean time, I'm going to remove the logs bit from this PR and merge this in.  OK with you @jacobtomlinson ?

Reply:
Sure :)